File: langs/java/util/src/main/java/dev/argon/nobleidl/util/CreateNobleIDLRuntimeOptions.java
package dev.argon.nobleidl.util;

import dev.argon.esexpr.ESExprBinaryWriter;
import dev.argon.esexpr.KeywordMapping;
import dev.argon.nobleidl.compiler.format.BackendMapping;
import dev.argon.nobleidl.compiler.format.BackendOptions;
import dev.argon.nobleidl.compiler.format.NobleIdlJarOptions;
import dev.argon.nobleidl.compiler.format.PackageMapping;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;

public class CreateNobleIDLRuntimeOptions {

	public static void main(String[] args) throws Throwable {
		var options = new NobleIdlJarOptions(
			List.of("dev/argon/nobleidl/runtime/nobleidl-core.nidl"),
			new BackendMapping(
				new KeywordMapping<>(Map.of(
					"java", new BackendOptions(
						new PackageMapping(
							new KeywordMapping<>(Map.of(
								"nobleidl.core", "dev.argon.nobleidl.runtime"
							))
						)
					)
				))
			)
		);

		try(var os = Files.newOutputStream(Path.of("../runtime/src/main/resources/nobleidl-options.esxb"))) {
			ESExprBinaryWriter.writeWithSymbolTable(os, NobleIdlJarOptions.codec().encode(options));
		}
	}

}

File: langs/scala/util/src/main/scala/nobleidl/util/CreateNobleIDLRuntimeOptions.scala
package nobleidl.util

import esexpr.{Dictionary, ESExprBinaryDecoder, ESExprBinaryEncoder, ESExprCodec}
import nobleidl.compiler.format.{BackendMapping, BackendOptions, NobleIdlJarOptions, PackageMapping}
import zio.stream.ZSink
import zio.{Scope, ZIO, ZIOAppArgs, ZIOAppDefault}

import java.nio.file.{Files, Path}

object CreateNobleIDLRuntimeOptions extends ZIOAppDefault {

  override def run: ZIO[ZIOAppArgs & Scope, Any, Any] =

    def buildOptions(backendMap: Map[String, BackendOptions]): NobleIdlJarOptions =
      NobleIdlJarOptions(
        idlFiles = Seq("nobleidl/core/nobleidl-core.nidl"),
        backends = BackendMapping(Dictionary(backendMap)),
      )

    val scalaBackends = Map(
      "scala" -> BackendOptions(
        packageMapping = PackageMapping(Dictionary(Map(
          "nobleidl.core" -> "nobleidl.core",
        ))),
      ),
    )

    val jvmBackends = scalaBackends
    val jsBackends = scalaBackends ++ Map(
      "scalajs" -> BackendOptions(
        packageMapping = PackageMapping(Dictionary(Map(
          "nobleidl.core" -> "nobleidl.sjs.core",
        ))),
      ),
    )

    def generate(platform: String, options: NobleIdlJarOptions): ZIO[Any, Throwable, Unit] =
      ESExprBinaryEncoder.writeWithSymbolTable(summon[ESExprCodec[NobleIdlJarOptions]].encode(options))
        .run(ZSink.fromPath(Path.of(s"runtime/$platform/src/main/resources/nobleidl-options.esxb").nn))
        .unit

    generate("jvm", buildOptions(jvmBackends)) &>
      generate("js", buildOptions(jsBackends))
  end run

}
